### PR TITLE
Fix reader copying by storing init arguments

### LINKED ISSUE #3664: `ProtoReader` based classes don't use `kwargs` on `copy`
Follow-up from #3642 

The `copy` method of `ProtoReader` derived classes doesn't account for kwargs which were passed on class construction. The use of `convert_units` was fixed in #3642 but we should do it for kwargs too.

### PR DESCRIPTION
**Pull request by IAlibay**
Fixes #3664

This is the final PR needed for 2.2.0.

So this isn't a very pretty fix but it's the best I could come up with for now. I'm happy to go with alternatives, including deprecating / removing `copy`.

Changes made in this Pull Request:
 - Added decorator for storing class initialisation arguments.
 - Fixed `copy` methods for readers, allowing for copies to get the same init arguments as the original object

Todo
 - Need to update a couple of tests for a few readers that have custom init arguments (just to check that everything works as intended since the base classes aren't used by all readers).

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by richardjgowers** [comment]
I like the decorator, but I think maybe `MemoryReader` is again going to be an exception.  If I create a `MemoryReader`, modify it's coordinates (i.e. unwrap or something) then copy it, I expect to get the same copied coordinates back.  I *think* this is what will happen with the current code (since it's all shallow copies).  One wrinkle is the docs say you'll get "Return independent copy of this Reader.
        New Reader will have its own file handle and can seek/iterate
        independently of the original".  This won't quite be true with memoryreader as the new Reader will share the same numpy array, so modifying one affects the other.  I think maybe memoryreader will have to initiate the copy with a deepcopy of its internal coordinates to behave more like other readers here.

**Comment by IAlibay** [comment]
> I like the decorator, but I think maybe `MemoryReader` is again going to be an exception. If I create a `MemoryReader`, modify it's coordinates (i.e. unwrap or something) then copy it, I expect to get the same copied coordinates back. I _think_ this is what will happen with the current code (since it's all shallow copies). One wrinkle is the docs say you'll get "Return independent copy of this Reader. New Reader will have its own file handle and can seek/iterate independently of the original". This won't quite be true with memoryreader as the new Reader will share the same numpy array, so modifying one affects the other. I think maybe memoryreader will have to initiate the copy with a deepcopy of its internal coordinates to behave more like other readers here.

Thanks for not shooting down the decorator 😅 @richardjgowers (although I still think it's ugly)

re: memoryreader - doesn't it already do deepcopies for coordinates/velocities/positions/dimensions? https://github.com/MDAnalysis/mdanalysis/blob/1945878e224beb1076be258432d911d5fda5f54a/package/MDAnalysis/coordinates/memory.py#L431-L437

**Comment by richardjgowers** [comment]
Ah ok cool, didn't realise MemoryReader was already clever.  So long as the input arrays aren't captured by the decorator this looks good

**Comment by IAlibay** [comment]
> So long as the input arrays aren't captured by the decorator this looks good

Yeah I got caught out by that early on whilst doing this fix 🤣 hence why for the memoryreader case I just do a shallow copy of the kwargs input dictionary only (so it should only be for `dt` and `time_offset`).

File: package/MDAnalysis/coordinates/DCD.py
Comment (lilyminium) [inline package/MDAnalysis/coordinates/DCD.py:72]: Are there readers that would not want to store init arguments? If not, would it be easier to use the Reader metaclass to wrap `__init__`? Or would that lead to problems subclassing

Comment (IAlibay) [reply package/MDAnalysis/coordinates/DCD.py:72]: MemoryReader is a good example of a case where we wouldn't want to store all the init arguments (you could but then it gets rather messy to extract out just the kwargs, and you also end up capturing references to all the input arrays in `self._kwargs`).

There is probably a very elegant way to do this using metaclasses, however I'm rather short on my self-allocated time to get 2.2.0 done. I gave it a quick shot and failed, so unless someone has a very specific "here is how I would do it" instruction, I might say that we kick this down the bucket as a cleanup issue for the future?

File: testsuite/MDAnalysisTests/coordinates/test_copying.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_copying.py:37]: 🤦🏽 I had completely forgot this set of tests existed (to be fair, looks like we've been forgetting about it for the last few readers too).

**Review by IAlibay — COMMENTED** [review]
Sorry for the mess here, ended up going back and forth on tests so there's quite a few files with just pep8 / import order fixes (didn't think it worth changing back).

Overall this should be it for now. Thanks for the comments up until now, if we could get a formal yeah/nay on this that would be great @MDAnalysis/coredevs, it is one of the last things holding up 2.2.0.

File: package/MDAnalysis/coordinates/memory.py
Comment (orbeckst) [inline package/MDAnalysis/coordinates/memory.py]:
```suggestion
            Input kwargs are now stored under the :attr:`_kwargs` attribute, and
```
or use double backticks.

File: package/MDAnalysis/lib/util.py
Comment (orbeckst) [inline package/MDAnalysis/lib/util.py]: Why is `shutil` added if it's not used in your code?

File: package/MDAnalysis/lib/util.py
Comment (orbeckst) [inline package/MDAnalysis/lib/util.py:2440]: Whenever I see `inspect` I shudder... :-p

File: testsuite/MDAnalysisTests/coordinates/test_copying.py
Comment (orbeckst) [reply testsuite/MDAnalysisTests/coordinates/test_copying.py:37]: You earned the test archaeologist badge 🏅

File: testsuite/MDAnalysisTests/coordinates/test_copying.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/coordinates/test_copying.py]:
```suggestion
    # memory and chain readers subclass protoreader directly and
```

File: testsuite/MDAnalysisTests/coordinates/test_mol2.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/coordinates/test_mol2.py]: Previously, the tests worked _without_ this line??!?

File: testsuite/MDAnalysisTests/coordinates/test_trj.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/coordinates/test_trj.py]: Previously, the tests worked without this line??!?

**Review by orbeckst — APPROVED** [review]
That was a lot of worked for a "small" issue! Thank you for doing it all.

I understand the rationale for using the decorator approach instead of introducing yet another "base/proto/intermediate" class in the readers. Maybe in the future we have a rethink on how to organize the readers... perhaps as part of the Cythonization effort?

I have minor comments/suggestions but nothing blocking.

**Comment by orbeckst** [comment]
There's an issue with 3.10, though
```
>       r = get_reader_for(filename, format=fmt_name)(filename, **extras)

/home/runner/work/mdanalysis/mdanalysis/testsuite/MDAnalysisTests/coordinates/test_copying.py:101: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/runner/work/mdanalysis/mdanalysis/package/MDAnalysis/lib/util.py:2460: in wrapper
    return func(self, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'ChemfilesReader' object has no attribute 'filename'") raised in repr()] ChemfilesReader object at 0x7fc5308527a0>
filename = '/home/runner/work/mdanalysis/mdanalysis/testsuite/MDAnalysisTests/data/lammps/datatest.data'
chemfiles_format = 'LAMMPS Data', kwargs = {}
```

**Comment by orbeckst** [comment]
... ah, you were fixing it while I was complaining...

File: testsuite/MDAnalysisTests/coordinates/test_chemfiles.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_chemfiles.py:105]: These extra tests in chemfiles and H5MD aren't super pretty, but it's the only way I could think of having them sit within import guarded regions without making `test_copying` overly complicated.

File: package/MDAnalysis/lib/util.py
Comment (IAlibay) [reply package/MDAnalysis/lib/util.py]: oops here and below are leftover imports from me testing out various things - should be fixed now.

**Comment by IAlibay** [comment]
Thanks for the review @orbeckst, hopefully the last few commits should have addressed your comments?

>  Maybe in the future we have a rethink on how to organize the readers... perhaps as part of the Cythonization effort?

Yeah 100%, I'll open an issue after this is merged with a general "re-organising readers. We probably should also one day get back to moving the reader tests to use the base classes too.

> ... ah, you were fixing it while I was complaining...

Yeah sorry I noticed that the readers with optional deps don't like being in `test_copying`. I've added individual tests for each of them, it's not pretty, but at least it should do a suitable job of checking things work for now?

File: testsuite/MDAnalysisTests/lib/test_util.py
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/lib/test_util.py:2159]: They're pretty delicious

**Review by richardjgowers — APPROVED** [review]

**Pull request merged**
